lib/ansible/plugins/connections/docker.py

```diff
@@ -0,0 +1,104 @@
+
+    def __init__(self, connection_info, new_stdin, *args, **kwargs):
+        super(Connection, self).__init__(connection_info, new_stdin, *args, **kwargs)
+
+        if 'docker_command' in kwargs:
+            self.docker_cmd = kwargs['docker_command']
+        else:
+            self.docker_cmd = 'docker'
+
+    @property
+    def transport(self):
+        return 'docker'
+
+    def _connect(self, port=None):
+        """ Connect to the container. Nothing to do """
+        return self
+
+    def exec_command(self, cmd, tmp_path, sudo_user=None, sudoable=False,
+                     executable='/bin/sh', in_data=None, su=None,
+                     su_user=None):
+
+        """ Run a command on the local host """
+
+        # Don't currently support su
+        if su or su_user:
+            raise errors.AnsibleError("Internal Error: this module does not "
+                                      "support running commands via su")
+
+        if in_data:
+            raise errors.AnsibleError("Internal Error: this module does not "
+                                      "support optimized module pipelining")
+
+        if sudoable and sudo_user:
+            raise errors.AnsibleError("Internal Error: this module does not "
+                                      "support running commands via sudo")
+
+        if executable:
+            local_cmd = [self.docker_cmd, "exec", self._connection_info.remote_addr, executable,
+                         '-c', cmd]
+        else:
+            local_cmd = '%s exec "%s" %s' % (self.docker_cmd, self._connection_info.remote_addr, cmd)
+
+        self._display.vvv("EXEC %s" % (local_cmd), host=self._connection_info.remote_addr)
+        p = subprocess.Popen(local_cmd,
+                             shell=isinstance(local_cmd, basestring),
+                             stdin=subprocess.PIPE,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        stdout, stderr = p.communicate()
+        return (p.returncode, '', stdout, stderr)
+
+    # Docker doesn't have native support for copying files into running
+    # containers, so we use docker exec to implement this
+    def put_file(self, in_path, out_path):
+        """ Transfer a file from local to container """
+        args = [self.docker_cmd, "exec", "-i", self._connection_info.remote_addr, "bash", "-c",
+                "cat > %s" % format(out_path)]
+
+        self._display.vvv("PUT %s TO %s" % (in_path, out_path), host='localhost')
+
+        if not os.path.exists(in_path):
+            raise errors.AnsibleFileNotFound(
+                "file or module does not exist: %s" % in_path)
+        p = subprocess.Popen(args, stdin=open(in_path),
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        p.communicate()
+
+        # HACK: Due to a race condition, this sometimes returns before
+        # the file has been written to disk, so we sleep for one second
+        time.sleep(1)
+
```

Comment:
@lorin Thanks for the elaboration, I'll keep an eye open for any issues that might be caused by this.

For now I've removed the sleep call. Marking the plugin as experimental might be a good idea as well, although the rest of the functionality seems to work well for me (although it hasn't been battle tested of course).

Thanks again for your attention on this.